### PR TITLE
DM-27478: Add direct URI transfer option

### PR DESCRIPTION
**Pull request by timj**

### CODE REVIEW

File: python/lsst/daf/butler/core/_butlerUri/_butlerUri.py [hunk 1]
```diff
@@ -382,6 +382,34 @@ def updateFile(self, newfile: str) -> None:
...
         self._uri = self._uri._replace(path=newpath)
 
+    def updateExtension(self, ext: Optional[str]) -> None:
+        """Update the file extension associated with this `ButlerURI` in place.
+
+        All file extensions are replaced.
+
+        Parameters
+        ----------
+        ext : `str`
```
Comment (andy-slac) [inline python/lsst/daf/butler/core/_butlerUri/_butlerUri.py]: `str` or `None`?

File: python/lsst/daf/butler/core/location.py [hunk 2]
```diff
@@ -43,60 +41,71 @@ class Location:
     ----------
     datastoreRootUri : `ButlerURI` or `str`
```
Comment (andy-slac) [inline python/lsst/daf/butler/core/location.py]: `ButlerURI` or `str` or `None`?

File: python/lsst/daf/butler/datastores/fileDatastore.py [hunk 3]
```diff
@@ -721,6 +731,12 @@ def _extractIngestInfo(self, path: Union[str, ButlerURI], ref: DatasetRef, *,
                     raise RuntimeError(f"Unexpectedly learned that {srcUri} is "
                                        f"not within datastore {self.root}")
                 tgtLocation = self.locationFactory.fromPath(pathInStore)
+        elif transfer == "direct":
+            # Want to store the full URI to the resource directly in
+            # datastore. This is useful for referring to permanent archive
+            # storage for raw data.
+            # Trust that people know what they are doing.
```
Comment (andy-slac) [inline python/lsst/daf/butler/datastores/fileDatastore.py:738]: As we always do 😄

**Review by andy-slac — APPROVED** [review]
Looks good to me, but I'm not a butler expert (or expert butler). Couple of comments for docstrings.

File: python/lsst/daf/butler/core/location.py (hunk 2 above)
Comment (timj) [reply python/lsst/daf/butler/core/location.py]: The problem of numpydoc and type annotations getting out of sync...

**Pull request merged**